File: backend/app/models/interview.py
import enum
import uuid

from sqlalchemy import (
    JSON,
    Column,
    DateTime,
    Enum,
    Float,
    ForeignKey,
    Integer,
    String,
    Text,
    UniqueConstraint,
    and_,
    or_,
)
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.db.session import Base


class InterviewType(str, enum.Enum):
    """Features 1-4: the four kinds of interview this module generates."""

    HR = "HR"
    TECHNICAL = "TECHNICAL"
    BEHAVIORAL = "BEHAVIORAL"
    APTITUDE = "APTITUDE"


class Difficulty(str, enum.Enum):
    """Feature 5."""

    EASY = "EASY"
    MEDIUM = "MEDIUM"
    HARD = "HARD"


class SessionStatus(str, enum.Enum):
    """
    Feature 7: where an interview sits in its lifecycle.

        CREATED ──start──> IN_PROGRESS ──end──> COMPLETED
                               │  ▲
                          pause│  │resume
                               ▼  │
                            PAUSED ──end──> COMPLETED

    ABANDONED is not a transition anyone requests — it is where a session that
    was never finished ends up. COMPLETED is terminal: an interview that has
    been ended cannot be restarted, because the questions have been seen.
    """

    CREATED = "CREATED"
    IN_PROGRESS = "IN_PROGRESS"
    PAUSED = "PAUSED"
    COMPLETED = "COMPLETED"
    ABANDONED = "ABANDONED"


# The states in which a candidate is partway through: a question may be served,
# and the interview is neither finished nor unstarted. Defined once because the
# REST endpoints, the voice socket and analytics all need the same answer.
ACTIVE_STATUSES = (SessionStatus.IN_PROGRESS, SessionStatus.PAUSED)


class QuestionSource(str, enum.Enum):
    """Whether the questions came from the AI or the built-in bank."""

    AI = "AI"
    FALLBACK = "FALLBACK"


class Interview(Base):
    __tablename__ = "interviews"

    id = Column(Integer, primary_key=True, index=True)

    # An opaque public identifier for this interview session.
    #
    # One interview is one session, so this identifies the same run as `id` —
    # but it is the one that is safe to put in a URL, a log line or a support
    # ticket. A sequential integer tells anyone who sees it roughly how many
    # interviews the platform has ever run, and invites guessing at neighbours;
    # a UUID tells them nothing.
    #
    # Generated in Python rather than by the database so it is available
    # immediately on an unflushed object, and so the SQLite fallback behaves
    # the same as Postgres.
    session_id = Column(
        String(36),
        nullable=False,
        unique=True,
        index=True,
        default=lambda: str(uuid.uuid4()),
    )

    user_id = Column(
        Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )

    interview_type = Column(Enum(InterviewType), nullable=False)

    # Feature 6: deliberately a free-text column, not an enum. A new domain is
    # just a new string — no migration and no code change. SUGGESTED_DOMAINS in
    # services/question_bank.py is only a hint list for the UI.
    domain = Column(String(120), nullable=False)

    difficulty = Column(Enum(Difficulty), nullable=False)

    # --- session management (feature 7) ---
    status = Column(
        Enum(SessionStatus), nullable=False, default=SessionStatus.CREATED, index=True
    )
    question_count = Column(Integer, nullable=False, default=0)
    source = Column(Enum(QuestionSource), nullable=False, default=QuestionSource.FALLBACK)

    started_at = Column(DateTime(timezone=True), nullable=True)
    completed_at = Column(DateTime(timezone=True), nullable=True)

    # When the current pause began. Null whenever the interview is not paused —
    # this is the open half of a pause interval, not a history of them.
    paused_at = Column(DateTime(timezone=True), nullable=True)

    # Total time spent paused, accumulated on each resume.
    #
    # Without this, pause would be a way to get unlimited thinking time while
    # the countdown quietly kept running, and the elapsed clock would report a
    # duration the candidate did not spend interviewing. Both numbers are meant
    # to describe real time at the keyboard, so paused time is subtracted
    # rather than ignored.
    total_paused_seconds = Column(Integer, nullable=False, default=0, server_default="0")

    # How long the interview actually took, in seconds, stamped when it ends.
    #
    # Derivable from completed_at - started_at - total_paused_seconds, and it
    # is derived exactly that way — but it is *stored* because it is a fact
    # about a finished interview, and a stored fact does not change if the
    # formula is later corrected or if paused time is adjusted. Null while the
    # interview is still running: there is no duration until there is an end.
    duration_seconds = Column(Integer, nullable=True)

    # Module 4: how long the candidate gets per question, in seconds.
    #
    # Snapshotted from PlatformSettings.session_minutes when the interview
    # starts rather than read live, so an administrator editing the platform
    # setting mid-session cannot move the goalposts under a candidate who is
    # already answering. Null on interviews created before timing existed —
    # those simply have no countdown.
    question_seconds = Column(Integer, nullable=True)

    # Module 6: the rubric score for this interview, 0-100 — the average of
    # its answered questions' scores (app.services.scoring.aggregate_score).
    # Stamped once when the interview reaches COMPLETED, the same "stored
    # fact about a finished interview" reasoning as duration_seconds above: it
    # must not silently reinterpret itself if the rubric weights are ever
    # retuned. Null while running, and null forever on an interview where no
    # answer was ever scored (no key, provider down, or every question
    # skipped) — that is a different fact from "scored zero" and must not be
    # displayed as one.
    overall_score = Column(Float, nullable=True)

    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )

    questions = relationship(
        "InterviewQuestion",
        back_populates="interview",
        cascade="all, delete-orphan",
        order_by="InterviewQuestion.sequence_no",
    )

    def __repr__(self) -> str:
        return (
            f"<Interview id={self.id} type={self.interview_type} "
            f"domain={self.domain!r} difficulty={self.difficulty}>"
        )


class InterviewQuestion(Base):
    __tablename__ = "interview_questions"

    # A question's position within its interview is unique — this is what keeps
    # the voice interviewer from asking the same question twice.
    __table_args__ = (
        UniqueConstraint("interview_id", "sequence_no", name="uq_interview_question_seq"),
    )

    id = Column(Integer, primary_key=True, index=True)
    interview_id = Column(
        Integer, ForeignKey("interviews.id", ondelete="CASCADE"), nullable=False, index=True
    )

    question_text = Column(Text, nullable=False)
    category = Column(String(80), nullable=False)
    difficulty = Column(Enum(Difficulty), nullable=False)
    sequence_no = Column(Integer, nullable=False)

    # --- filled in by the voice interviewer (feature 9) ---
    # Written answers are not collected today. The column stays because rows
    # created before answers became audio still hold transcripts here, and a
    # future scoring module will want somewhere to put a written answer.
    answer_text = Column(Text, nullable=True)
    # The candidate's spoken answer is stored as the recording itself — the
    # platform no longer transcribes it. Path is relative to ANSWER_AUDIO_DIR's
    # parent and is always a name the server chose, never a client filename.
    answer_audio_path = Column(String(512), nullable=True)
    answer_audio_mime = Column(String(80), nullable=True)
    # How long the candidate actually spoke for, in seconds, as measured by the
    # browser's recorder.
    #
    # Deliberately NOT derived from answered_at - asked_at. That interval also
    # contains thinking time, re-reading the question and hesitation before
    # hitting record. Words-per-minute computed against it would be wrong in a
    # way that always flatters nobody and penalises the careful, so pace uses
    # this measured speaking time or it is not reported at all.
    answer_duration_seconds = Column(Float, nullable=True)

    # Module 5 output for this one answer: filler counts, pace, grammar notes,
    # communication assessment. JSON rather than columns because the shape will
    # churn as the analysis is tuned — same reasoning as the résumé extraction.
    # Null means not analysed, which is not the same as "analysed and clean".
    analysis = Column(JSON, nullable=True)
    analyzed_at = Column(DateTime(timezone=True), nullable=True)

    asked_at = Column(DateTime(timezone=True), nullable=True)
    answered_at = Column(DateTime(timezone=True), nullable=True)
    # Set when the candidate passes on a question. A skip is *not attempted* —
    # deliberately a separate column from answered_at rather than an empty
    # answer, so "said nothing" can never be counted as "answered".
    skipped_at = Column(DateTime(timezone=True), nullable=True)

    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    interview = relationship("Interview", back_populates="questions")

    @property
    def time_on_question_seconds(self):
        """
        Seconds from being asked to being answered or skipped.

        A property rather than a column: it is fully determined by timestamps
        already stored, so persisting it would create a second copy that could
        drift. Lives here so the response schemas pick it up through
        from_attributes without every endpoint recomputing it.

        Imported lazily to keep the model layer free of a service import at
        module scope.
        """
        from app.services.session_timing import question_seconds_spent

        return question_seconds_spent(self)

    def __repr__(self) -> str:
        return f"<InterviewQuestion id={self.id} interview_id={self.interview_id} seq={self.sequence_no}>"


# One definition of each state, shared by the interviewer's progress counter and
# by every analytics endpoint, so the two can never disagree.
#
# Both expressions also read rows written before answers became audio, where a
# transcript in answer_text meant answered and an empty string meant skipped.

# Attempted: the candidate actually said something.
QUESTION_ANSWERED = or_(
    InterviewQuestion.answer_audio_path.isnot(None),
    and_(InterviewQuestion.answer_text.isnot(None), InterviewQuestion.answer_text != ""),
)

# Not attempted: the candidate passed on it.
#
# The IS NOT NULL guard is load-bearing. In SQL, `answer_text = ''` against a
# NULL column evaluates to NULL rather than false, and a NULL here propagates
# through OR into QUESTION_HANDLED — whose negation is then also NULL, so an
# untouched question would match nothing and the interview would look finished
# before it began.
QUESTION_SKIPPED = or_(
    InterviewQuestion.skipped_at.isnot(None),
    and_(InterviewQuestion.answer_text.isnot(None), InterviewQuestion.answer_text == ""),
)

# Dealt with either way — answered or skipped. This is what decides whether the
# interviewer moves on, so a skipped question is never asked twice.
QUESTION_HANDLED = or_(QUESTION_ANSWERED, QUESTION_SKIPPED)

File: backend/models/Interview.js
const pool = require('../db');

class Interview {
  // Create a new interview session
  static async create({ user_id, interview_type, domain, difficulty, questions }) {
    try {
      const result = await pool.query(
        `INSERT INTO interviews (user_id, interview_type, domain, difficulty, questions, status, created_at)
         VALUES ($1, $2, $3, $4, $5, 'pending', CURRENT_TIMESTAMP)
         RETURNING *`,
        [user_id, interview_type, domain, difficulty, JSON.stringify(questions)]
      );
      console.log('✅ Interview created:', result.rows[0].id);
      return result.rows[0];
    } catch (error) {
      console.error('❌ Error creating interview:', error);
      throw error;
    }
  }

  // Get all interviews for a user
  static async findByUserId(user_id) {
    try {
      const result = await pool.query(
        `SELECT * FROM interviews WHERE user_id = $1 ORDER BY created_at DESC`,
        [user_id]
      );
      return result.rows;
    } catch (error) {
      console.error('❌ Error finding interviews:', error);
      throw error;
    }
  }

  // Get interview by ID
  static async findById(id) {
    try {
      const result = await pool.query(
        'SELECT * FROM interviews WHERE id = $1',
        [id]
      );
      return result.rows[0];
    } catch (error) {
      console.error('❌ Error finding interview:', error);
      throw error;
    }
  }

  // Update interview status with timestamps
  static async updateStatus(id, status) {
    try {
      let query = `
        UPDATE interviews 
        SET status = $1, updated_at = CURRENT_TIMESTAMP
      `;
      const params = [status, id];
      
      if (status === 'in_progress') {
        query = `
          UPDATE interviews 
          SET status = $1, 
              start_time = COALESCE(start_time, CURRENT_TIMESTAMP),
              updated_at = CURRENT_TIMESTAMP
          WHERE id = $2
          RETURNING *
        `;
      }
      else if (status === 'completed' || status === 'ended') {
        query = `
          UPDATE interviews 
          SET status = $1, 
              end_time = CURRENT_TIMESTAMP,
              updated_at = CURRENT_TIMESTAMP
          WHERE id = $2
          RETURNING *
        `;
      }
      else if (status === 'paused') {
        query = `
          UPDATE interviews 
          SET status = $1, 
              updated_at = CURRENT_TIMESTAMP 
          WHERE id = $2
          RETURNING *
        `;
      }
      
      const result = await pool.query(query, params);
      console.log(`✅ Interview ${id} status updated to: ${status}`);
      return result.rows[0];
    } catch (error) {
      console.error('❌ Error updating interview status:', error);
      throw error;
    }
  }

  // =============================================
  // UPDATE SCORE AND FEEDBACK - WITH SUBMISSION TYPE
  // =============================================
  static async updateScoreAndFeedback(id, score, feedback, submissionType = 'full', speechAnalysis = null) {
  try {
    const numericScore = parseFloat(score) || 0;
    
    const result = await pool.query(
      `UPDATE interviews 
       SET score = $1, 
           feedback = $2, 
           status = 'completed', 
           submission_type = $3,
           speech_analysis = $4,  -- ✅ Must have this column
           end_time = CURRENT_TIMESTAMP,
           updated_at = CURRENT_TIMESTAMP 
       WHERE id = $5 
       RETURNING *`,
      [numericScore, feedback, submissionType, speechAnalysis ? JSON.stringify(speechAnalysis) : null, id]
    );
    return result.rows[0];
  } catch (error) {
    console.error('❌ Error updating interview:', error);
    throw error;
  }
}

  // Pause interview session
  static async pauseInterview(id) {
    try {
      const result = await pool.query(
        `UPDATE interviews 
         SET status = 'paused', 
             updated_at = CURRENT_TIMESTAMP 
         WHERE id = $1 
         RETURNING *`,
        [id]
      );
      console.log(`⏸️ Interview ${id} paused`);
      return result.rows[0];
    } catch (error) {
      console.error('❌ Error pausing interview:', error);
      throw error;
    }
  }

  // Resume interview session
  static async resumeInterview(id) {
    try {
      const result = await pool.query(
        `UPDATE interviews 
         SET status = 'in_progress', 
             updated_at = CURRENT_TIMESTAMP 
         WHERE id = $1 
         RETURNING *`,
        [id]
      );
      console.log(`▶️ Interview ${id} resumed`);
      return result.rows[0];
    } catch (error) {
      console.error('❌ Error resuming interview:', error);
      throw error;
    }
  }

  // Delete interview
  static async delete(id) {
    try {
      const result = await pool.query(
        'DELETE FROM interviews WHERE id = $1 RETURNING *',
        [id]
      );
      console.log(`🗑️ Interview ${id} deleted`);
      return result.rows[0];
    } catch (error) {
      console.error('❌ Error deleting interview:', error);
      throw error;
    }
  }

  // Get active session for user
  static async getActiveSession(user_id) {
    try {
      const result = await pool.query(
        `SELECT * FROM interviews 
         WHERE user_id = $1 
         AND status IN ('pending', 'in_progress', 'paused')
         ORDER BY created_at DESC 
         LIMIT 1`,
        [user_id]
      );
      return result.rows[0] || null;
    } catch (error) {
      console.error('❌ Error getting active session:', error);
      throw error;
    }
  }

  // Get session stats for user
  static async getSessionStats(user_id) {
    try {
      const result = await pool.query(
        `SELECT 
          COUNT(*) as total_sessions,
          COUNT(CASE WHEN status = 'completed' THEN 1 END) as completed_sessions,
          COUNT(CASE WHEN status = 'in_progress' THEN 1 END) as active_sessions,
          COUNT(CASE WHEN status = 'paused' THEN 1 END) as paused_sessions,
          AVG(score) as average_score,
          MAX(score) as highest_score,
          MIN(score) as lowest_score,
          SUM(CASE WHEN status = 'completed' THEN EXTRACT(EPOCH FROM (end_time - start_time)) ELSE 0 END) as total_time_spent
         FROM interviews 
         WHERE user_id = $1`,
        [user_id]
      );
      return result.rows[0];
    } catch (error) {
      console.error('❌ Error getting session stats:', error);
      throw error;
    }
  }

  // Get interviews by type and domain
  static async findByTypeAndDomain(user_id, interview_type, domain) {
    try {
      const result = await pool.query(
        `SELECT * FROM interviews 
         WHERE user_id = $1 
         AND interview_type = $2 
         AND domain = $3 
         ORDER BY created_at DESC`,
        [user_id, interview_type, domain]
      );
      return result.rows;
    } catch (error) {
      console.error('❌ Error finding interviews by type and domain:', error);
      throw error;
    }
  }

  // =============================================
  // DETAILED SCORING METHODS
  // =============================================

  /**
   * Store detailed score breakdown
   */
  static async updateDetailedScores(id, scoreBreakdown) {
    try {
      const result = await pool.query(
        `UPDATE interviews 
         SET communication_score = $1,
             confidence_score = $2,
             technical_score = $3,
             professionalism_score = $4,
             score_breakdown = $5,
             updated_at = CURRENT_TIMESTAMP
         WHERE id = $6 
         RETURNING *`,
        [
          scoreBreakdown.communication || 0,
          scoreBreakdown.confidence || 0,
          scoreBreakdown.technical || 0,
          scoreBreakdown.professionalism || 0,
          JSON.stringify(scoreBreakdown),
          id
        ]
      );
      console.log('✅ Detailed scores updated for interview:', id);
      return result.rows[0];
    } catch (error) {
      // Column might not exist yet, gracefully handle
      if (error.message.includes('column') && error.message.includes('does not exist')) {
        console.warn('⚠️  Score columns not yet in database. This is OK - scores will be stored in feedback JSON');
        return null;
      }
      console.error('❌ Error updating detailed scores:', error);
      throw error;
    }
  }

  /**
   * Get interview with detailed scoring information
   */
  static async findByIdWithScores(id) {
    try {
      const result = await pool.query(
        `SELECT * FROM interviews WHERE id = $1`,
        [id]
      );
      
      if (result.rows[0]) {
        const interview = result.rows[0];
        
        // Parse feedback if it contains scoring data
        if (interview.feedback) {
          try {
            interview.feedbackData = JSON.parse(interview.feedback);
          } catch (e) {
            interview.feedbackData = null;
          }
        }
        
        return interview;
      }
      
      return null;
    } catch (error) {
      console.error('❌ Error finding interview with scores:', error);
      throw error;
    }
  }

  /**
   * Store AI-generated feedback and performance report
   */
  static async updateAIFeedback(id, feedbackData) {
    try {
      const result = await pool.query(
        `UPDATE interviews 
         SET feedback = $1,
             updated_at = CURRENT_TIMESTAMP
         WHERE id = $2 
         RETURNING *`,
        [JSON.stringify(feedbackData), id]
      );
      console.log('✅ AI feedback updated for interview:', id);
      return result.rows[0];
    } catch (error) {
      console.error('❌ Error updating AI feedback:', error);
      throw error;
    }
  }

  /**
   * Get performance metrics for user
   */
  static async getUserPerformanceMetrics(user_id) {
    try {
      const result = await pool.query(
        `SELECT 
          COUNT(*) as total_interviews,
          COUNT(CASE WHEN status = 'completed' THEN 1 END) as completed_interviews,
          AVG(score) as average_score,
          MAX(score) as highest_score,
          MIN(score) as lowest_score,
          COUNT(CASE WHEN score >= 90 THEN 1 END) as excellent_count,
          COUNT(CASE WHEN score >= 75 AND score < 90 THEN 1 END) as good_count,
          COUNT(CASE WHEN score >= 60 AND score < 75 THEN 1 END) as average_count,
          COUNT(CASE WHEN score >= 40 AND score < 60 THEN 1 END) as needs_improvement_count,
          COUNT(CASE WHEN score < 40 THEN 1 END) as poor_count
         FROM interviews 
         WHERE user_id = $1 AND status = 'completed'`,
        [user_id]
      );
      return result.rows[0];
    } catch (error) {
      console.error('❌ Error getting performance metrics:', error);
      throw error;
    }
  }

  /**
   * Get interview score history for trending
   */
  static async getScoreHistory(user_id, limit = 10) {
    try {
      const result = await pool.query(
        `SELECT 
          id,
          interview_type,
          domain,
          score,
          created_at,
          status
         FROM interviews 
         WHERE user_id = $1 AND status = 'completed'
         ORDER BY created_at DESC
         LIMIT $2`,
        [user_id, limit]
      );
      return result.rows.reverse(); // Return in ascending order for trending
    } catch (error) {
      console.error('❌ Error getting score history:', error);
      throw error;
    }
  }
}

module.exports = Interview;